File: src/client/Bridges/GameBridge.cpp
#include "GameBridge.hpp"

void GameBridge::callJoinGame(const QString& ip, const int& port, const QString& gameId,
                              const Role& role) {
  qDebug() << "GameBridge::callJoinGame called with" << ip << port << gameId;
  emit requestJoinToGame(ip.toStdString(), port, gameId.toStdString(), role);
}

void GameBridge::callMatch(const QString& turnId, const int& symbolId) {
  qDebug() << "GameBridge::callMatch called with" << turnId << symbolId;
  emit requestMatch(turnId.toStdString(), symbolId);
}

void GameBridge::callQuitGame() {
  qDebug() << "GameBridge::callQuitGame called";
  emit requestQuitGame();
}

File: ui/src/store/gameBridge.ts
import type { QtSignal } from "@/bridge/channel";
import { qwebchannelInitializer } from "@/bridge/initialize";
import type {
  CommunicationStatus,
  ConnectionError,
  ConnectionStatus,
  GameInfo,
  MatchResult,
  Role,
} from "@/types/dobble";
import { mainStore } from ".";

export interface GameBridge {
  onServerConnectionStateChanged: QtSignal<(status: ConnectionStatus) => void>;
  onServerConnectionErrorOccured: QtSignal<(error: ConnectionError) => void>;
  onServerCommunicationStateChanged: QtSignal<(status: CommunicationStatus) => void>;

  onGameInfoChanged: QtSignal<(gameInfo: GameInfo) => void>;
  onMatchResult: QtSignal<(result: MatchResult) => void>;
  onGameQuit: QtSignal<() => void>;

  // slots
  callJoinGame(ip: string, port: number, gameId: string, role: Role): void;
  callStartGame: () => void;
  callMatch: (turnId: string, symbolId: number) => void;
  callQuitGame: () => void;
}

qwebchannelInitializer.onReady("game", (bridge) => {
  bridge.onGameInfoChanged.connect((gameInfo: GameInfo) => {
    mainStore.setState((state) => ({ game: { ...state.game, gameInfo } }));
  });

  bridge.onMatchResult.connect((result: MatchResult) => {
    mainStore.getState().game.setMatchResult(result);
  });

  bridge.onGameQuit.connect(() => {
    mainStore.setState((state) => ({ game: { ...state.game, gameInfo: null } }));
  });
});